stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,250 @@
+        adj = G.to_adjacency_matrix().tocoo()
+
+        # Function to map node IDs to indices for quicker node index lookups
+        self._node_lookup = G._get_index_for_nodes
+
+        degree_mat = diags(np.asarray(adj.sum(1)).ravel())
+        laplacian = degree_mat - adj
+
+        # FIXME(#853): add in option to compute wavelet transform using Chebysev polynomials
+
+        self.scales = np.array(scales).astype(np.float32)
+
+        if num_eigenvecs is None:
+            eigen_vals, eigen_vecs = self._sufficiently_sampled_eigs(
+                laplacian,
+                self.scales.min(),
+                min_delta=min_delta,
+                initial_sigma=self._INITIAL_EIGS_SIGMA,
+            )
+        else:
+            eigen_vals, eigen_vecs = eigs(
+                laplacian, k=num_eigenvecs, sigma=self._INITIAL_EIGS_SIGMA
+            )
+            eigen_vals = np.real(eigen_vals).astype(np.float32)
+
+        self.eigen_vecs = np.real(eigen_vecs).astype(np.float32)
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of U exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+
+        # a list of [U exp(-scale * eigenvalues) for scale in scales]
+        Ues = [self.eigen_vecs * np.exp(-s * eigen_vals) for s in scales]
+        self.Ues = tf.convert_to_tensor(np.stack(Ues, axis=0))
+
+    @staticmethod
+    def _sufficiently_sampled_eigs(laplacian, min_scale, min_delta, initial_sigma):
+        """
+        This function calculates increasing numbers of eigenvalues using a binary search until a sufficient number of
+        eigenvalues have been found to ensure accurate results of the GraphWave algorithm.
+
+        Args:
+            laplacian: the un-normalized graph laplacian.
+            min_scale: the smallest wavelet scale used.
+            min_delta: the minimum relative change in the eigenvalue norm. A relative change less than this stops
+                the eigenvalue search.
+        Returns:
+            eigen_vals: the eigenvalues found.
+            eigen_vecs: the eigenvectors found.
+        """
+        max_num_eigs = laplacian.shape[0] - 2
+
+        prev_eig_norm = 0.0
+        eig_max = initial_sigma
+        k = min(16, max_num_eigs)
+        eigen_vals = np.array([], dtype=np.float32)
+        eigen_vecs = []
+
+        # use increasing k (doubling per iter) until the filtered eigenvalue l2 norm
+        # increases by less than 10% since the last iter.
+        while True:
+            # if sigma=eigenvalue eigs will throw an error
+            new_eigen_vals, new_eigen_vecs = eigs(laplacian, k=k, sigma=eig_max + 1e-7)
```

Comment:
specifying the initial search vector `v0` `eigs(..., v0=v0`) should make this deterministic, but experiments on cora show that this is not the case. This could be a numerically stability issue. I'll investigate a bit further, but if that can't be resolved I think it'll be best to just go ahead with the Chebyshev approach which I've got working now.

Reply:
@huon I ran some experiments, and the number of `nan` eigenvalues/vectors returned is equal to the number of zero eigenvalues missed. The number of zero eigenvalues missed seems to be random despite initialisation, causing the `eig_norm` to be somewhat random and therefore everything else. 

The only solution I can think of is to restrict GraphWave to use graphs with 1 connected component so it only has 1 zero eigenvalue. (i think a fairly common constraint for Graph ML) I'm pretty sure the shift invert approach only finds interior eigenvalues, so will skip the lowest and never attempt to find a zero eigenvalue. 

This approach isn't ideal, and I think would require adding a method to StellarGraph to find the largest connected component. maybe its best to just focus on the Chebyshev method for now, what do you think?